### PR TITLE
Adds TestViewOps, updates documentation

### PR DESCRIPTION
**Pull request by mruberry**
Understanding which ops return views and which return tensors with new storage is a common user issue, and an issue for developers connecting accelerators to PyTorch, too. This generic test suite verifies that ops which should return views do (and a few ops that shouldn't don't).  The documentation has also been updated for .t(), permute(), unfold(), and select() to clarify they return views.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks !

Maybe for test readability (and making sure that they are properly registered as views in for autograd purposes), you could use a function like (no need for inputs to require gradients):
```python
def is_view(base, other):
  # We should properly mark them as views
  if not other._is_view() or other._base is not base:
    return False
  # They should actually point to the same memory
  if base.storage().data_ptr() != other.storage().data_ptr():
    return False
  return True
```

File: test/test_torch.py [hunk 1]
```diff
@@ -14628,6 +14628,144 @@ def test_ones_like_multiple_device(self, devices):
...
+
+    def test_contiguous_nonview(self, device):
+        t = torch.ones(5, 5)
+        v = t.t().contiguous()
+        v[0, 0] = 0
+        self.assertNotEqual(t[0, 0], v[0, 0])
+
+    def test_reshape_view(self, device):
+        t = torch.ones(5, 5)
+        v = torch.reshape(t, (25,))
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py]: The docs are clear that users should not RELY on this behavior, that's true, but the test is intended for accelerators who are implementing our API, and they need to be consistent with its behavior (even if that behavior is subject to change).

File: torch/_tensor_docs.py [hunk 2]
```diff
@@ -3110,7 +3110,7 @@ def callable(a, b) -> number
                r"""
 unfold(dimension, size, step) -> Tensor
 
-Returns a tensor which contains all slices of size :attr:`size` from
+Returns a view of the original tensor which contains all slices of size :attr:`size` from
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_tensor_docs.py]: Natalia and I were discussing this, too, and we also thought that most end users would be confused by views. There does not seem to be a great description of them in the current docs. Ailing has a PR, https://github.com/pytorch/pytorch/pull/32560, that ameliorates the situation, but we could do better. Let's think about expanding #32560 or creating a new PR?

File: test/test_torch.py [hunk 3]
```diff
@@ -14628,6 +14628,144 @@ def test_ones_like_multiple_device(self, devices):
         self.assertEqual(output, expected)
 
 
+# Tests that ops which should return views do return views
+class TestViewOps(TestCase):
+
+    def test_diagonal_view(self, device):
+        t = torch.ones((5, 5))
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py]: Yep, just a mistake in this version of the PR. I'm going to fix it when I add a test or two for the bracket operator (to avoid hitting the CI again).

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 700db188:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=700db18819276dc3db7ffe4d7d4573239978db9c).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4360040) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92046814) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92046814))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in README.md 
Auto-merging README.md 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/win-test-helpers/test_python_nn.bat 
Auto-merging .jenkins/pytorch/win-test-helpers/test_python_nn.bat 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/win-test-helpers/test_python_all_except_nn.bat 
Auto-merging .jenkins/pytorch/win-test-helpers/test_python_all_except_nn.bat 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/test.sh 
Auto-merging .jenkins/pytorch/test.sh 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/multigpu-test.sh 
Auto-merging .jenkins/pytorch/multigpu-test.sh 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/test_torch.py [hunk 4]
```diff
@@ -14662,6 +14662,180 @@ def test_ones_like_multiple_device(self, devices):
         self.assertEqual(output, expected)
 
 
+# Tests ops and indexing to ensure they return views (and new tensors) as
+# appropriate.
+# Note: these tests do not verify tensor storage is the same or different
```
Comment (albanD) [inline test/test_torch.py]: Sounds good.
They should still populate the `._is_view()` and `._base` no?

Comment (mruberry) [reply test/test_torch.py]: Aha! I didn't even realize we had ._is_view(), and it turns out XLA tensors do implement it! That's much better.

Comment (albanD) [reply test/test_torch.py]: But I would argue you want both the `._is_view()` AND the check that it changes the value inplace. We have history of doing one but not the other and it is **really** hard to fix afterwards.

Comment (mruberry) [reply test/test_torch.py]: Interesting. So you think we extend each with an ._is_view() check? Sounds good.

File: test/test_torch.py [hunk 5]
```diff
@@ -14662,6 +14662,263 @@ def test_ones_like_multiple_device(self, devices):
         self.assertEqual(output, expected)
 
 
+# Tests ops and indexing to ensure they return views (and new tensors) as
```
Comment (albanD) [inline test/test_torch.py]: Ho that's a very good point. We want that `base is not other` right (maybe add it to the `is_view_of()` function)? If I do `b = a.view_as(a)`, I want `b` to be a brand new Tensor.
But that can be added later.

File: test/test_torch.py [hunk 6]
```diff
@@ -14662,6 +14662,263 @@ def test_ones_like_multiple_device(self, devices):
...
+        if base.device.type == 'cpu' or base.device.type == 'cuda':
+            if base.storage().data_ptr() != other.storage().data_ptr():
+                return False
+
+        return True
+
+    def test_diagonal_view(self, device):
+        t = torch.ones((5, 5), device=device)
+        v = torch.diagonal(t)
+        assert self.is_view_of(t, v)
```
Comment (albanD) [inline test/test_torch.py]: nit: `self.assertTrue()` for all functions.

**Review by albanD — APPROVED** [review]
LGTM small nits only
